File: Fourth Semester - Design and Programming Environments/C#/ProiectC#/service/Service.cs
using log4net;
using TravelAgency.exception;
using TravelAgency.model;
using TravelAgency.repository;
using TravelAgency.utils;

namespace TravelAgency.service;

public class Service 
{
    private static readonly ILog log =
        LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
    
    private IAgentRepository _agentRepository;
    private IReservationRepository reservationRepository;
    private ITripRepository tripRepository;
    
    public Service(IAgentRepository agentRepository,
        IReservationRepository reservationRepository, ITripRepository tripRepository)
    {
        log.Info("Creating Service");
        this._agentRepository = agentRepository;
        this.reservationRepository = reservationRepository;
        this.tripRepository = tripRepository;
    }

    public Agent AddAgent(Agent agent)
    {
        log.Info("Adding agent " + agent);
        try
        {
            Agent result = _agentRepository.Save(agent) ?? throw new ServiceException("Failed to save agent");
            log.Info("Agent added successfully " + agent);
            return result;
        }
        catch (RepositoryException e)
        {
            log.Error("Failed to add agent " + agent, e);
            throw new ServiceException("Failed to add agent", e);
        }
    }

    public Agent SearchAgentByUsername(string username, string password)
    {
        log.Info("Searching agent by username " + username);
        try
        {
            Agent result = _agentRepository.FindByUsername(username, password) ??
                           throw new ServiceException("Failed to find agent by username");
            log.Info("Agent found " + result);
            return result;
        }
        catch (RepositoryException e)
        {
            log.Error("Failed to search agent by username " + username, e);
            throw new ServiceException("Failed to search agent", e);
        }
    }

    public Trip UpdateSeatsOfTrip(int id, int seatsReserved)
    {
        log.Info("Updating seats of trip " + id);
        try
        {
            Trip result = tripRepository.UpdateSeats(id, seatsReserved) ?? throw new ServiceException("Failed to update trip");
            log.Info("Trip updated successfully " + result);
            return result;
        }
        catch (RepositoryException e)
        {
            log.Error("Failed to update seats of trip " + id, e);
            throw new ServiceException("Failed to update seats of trip", e);
        }
    }

    public Reservation AddReservation(Reservation reservation)
    {
        log.Info("Adding reservation " + reservation);
        try
        {
            Reservation result = reservationRepository.Save(reservation) ?? throw new ServiceException("Failed to save reservation");
            Trip trip = UpdateSeatsOfTrip(reservation.Trip.Id, reservation.NumberOfReservedSeats);
            log.Info("Reservation added successfully " + reservation);
            return result;
        }
        catch (RepositoryException e)
        {
            log.Error("Failed to add reservation " + reservation, e);
            throw new ServiceException("Failed to add reservation", e);
        }
    }

    public Reservation DeleteReservation(int id)
    {
        log.Info("Deleting reservation " + id);
        try
        {
            Reservation result = reservationRepository.Delete(id) ?? throw new ServiceException("Failed to delete reservation");
            Trip trip = UpdateSeatsOfTrip(result.Trip.Id, -result.NumberOfReservedSeats);
            log.Info("Reservation deleted successfully " + result);
            return result;
        }
        catch (RepositoryException e)
        {
            log.Error("Failed to delete reservation " + id, e);
            throw new ServiceException("Failed to delete reservation", e);
        }
    }
    
    public List<Trip> GetAllTrips()
    {
        log.Info("Getting all trips");
        try
        {
            IEnumerable<Trip> result = tripRepository.FindAll() ?? throw new ServiceException("Failed to get all trips");
            log.Info("Trips retrieved successfully " + result);
            List<Trip> trips = result.ToList();
            return trips;
        }
        catch (RepositoryException e)
        {
            log.Error("Failed to get all trips", e);
            throw new ServiceException("Failed to get all trips", e);
        }
    }

    public List<Trip> SearchTripsByDestTime(string destination, DateTime departureDate, TimeSpan startTime,
        TimeSpan endTime)
    {
        log.Info("Searching trips by destination " + destination);
        try
        {
            IEnumerable<Trip> result = tripRepository.FindTripsByDestination(destination) ?? throw new ServiceException("Failed to search trips by destination");
            log.Info("Trips found " + result);
            List<Trip> trips = result.ToList();
            trips = trips.Where(t => t.DepartureDate.Date == departureDate.Date &&
                                     t.DepartureDate.TimeOfDay >= startTime &&
                                     t.DepartureDate.TimeOfDay <= endTime).ToList();
            return trips;
        }
        catch (RepositoryException e)
        {
            log.Error("Failed to search trips by destination " + destination, e);
            throw new ServiceException("Failed to search trips by destination", e);
        }
    }
    
    public List<Reservation> GetAllReservations()
    {
        log.Info("Getting all reservations");
        try
        {
            IEnumerable<Reservation> result = reservationRepository.FindAll() ?? throw new ServiceException("Failed to get all reservations");
            log.Info("Reservations retrieved successfully " + result);
            List<Reservation> reservations = result.ToList();
            return reservations;
        }
        catch (RepositoryException e)
        {
            log.Error("Failed to get all reservations", e);
            throw new ServiceException("Failed to get all reservations", e);
        }
    }
    
    public Trip GetTripById(int id)
    {
        log.Info("Getting trip by id " + id);
        try
        {
            Trip result = tripRepository.FindOne(id) ?? throw new ServiceException("Failed to get trip by id");
            log.Info("Trip retrieved successfully " + result);
            return result;
        }
        catch (RepositoryException e)
        {
            log.Error("Failed to get trip by id " + id, e);
            throw new ServiceException("Failed to get trip by id", e);
        }
    }
}

File: Fourth Semester - Design and Programming Environments/Java/ProiectJava/java/org/example/service/Service.java
package org.example.service;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.exception.RepositoryException;
import org.example.exception.ServiceException;
import org.example.modul.Agent;
import org.example.modul.Reservation;
import org.example.modul.Trip;
import org.example.repository.interfaces.AgentRepository;
import org.example.repository.interfaces.ReservationRepository;
import org.example.repository.interfaces.TripRepository;
import org.example.utils.event.ChangeEventType;
import org.example.utils.event.TravelAgencyEventType;
import org.example.utils.observer.Observable;
import org.example.utils.observer.Observer;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class Service implements Observable<TravelAgencyEventType> {
    private static final Logger logger = LogManager.getLogger();

    private final AgentRepository agentRepository;
    private final ReservationRepository reservationRepository;
    private final TripRepository tripRepository;
    private List<Observer<TravelAgencyEventType>> observers = new ArrayList<>();


    public Service(AgentRepository agentRepository, ReservationRepository reservationRepository, TripRepository tripRepository) {
        this.agentRepository = agentRepository;
        this.reservationRepository = reservationRepository;
        this.tripRepository = tripRepository;
    }

    public Agent addAgent(Agent agent) {
        logger.traceEntry("Adding agent {}", agent);
        try {
            Agent result = agentRepository.save(agent).orElseThrow(() -> new ServiceException("Failed to save agent"));
            logger.traceExit(result);
            return result;
        } catch (RepositoryException e) {
            logger.error("Error in addAgent: {}", e.getMessage());
            throw new ServiceException("Service error while adding agent", e);
        }
    }

    public Agent searchAgentByUsername(String username, String password){
        logger.traceEntry("Searching agent by username {}", username);
        try {
            Agent result = agentRepository.findByUsername(username, password).orElseThrow(() -> new ServiceException("Agent not found"));
            logger.traceExit(result);
            return result;
        } catch (RepositoryException e) {
            logger.error("Error in searchAgentByUsername: {}", e.getMessage());
            throw new ServiceException("Service error while searching agent by username", e);
        }
    }

    public Trip updateSeatsOfTrip(Integer id, Integer seatsReserved) {
        logger.traceEntry("Updating seats of trip with id {}", id);
        try {
            Trip result = tripRepository.updateSeats(id, seatsReserved).orElseThrow(() -> new ServiceException("Failed to update seats of trip"));
            logger.traceExit(result);
            notifyObservers(new TravelAgencyEventType<>(ChangeEventType.RESERVEDSEAT, result));
            return result;
        } catch (RepositoryException e) {
            logger.error("Error in updateSeatsOfTrip: {}", e.getMessage());
            throw new ServiceException("Service error while updating seats of trip", e);
        }
    }

    public Reservation addReservation(Reservation reservation) {
        logger.traceEntry("Adding reservation {}", reservation);
        try {
            Reservation result = reservationRepository.save(reservation).orElseThrow(() -> new ServiceException("Failed to save reservation"));
            Trip trip = updateSeatsOfTrip(reservation.getTrip().getId(), reservation.getNumberOfReservedSeats());
            logger.traceExit(result);
            notifyObservers(new TravelAgencyEventType<>(ChangeEventType.RESERVEDSEAT, result));
            return result;
        } catch (RepositoryException e) {
            logger.error("Error in addReservation: {}", e.getMessage());
            throw new ServiceException("Service error while adding reservation", e);
        }
    }

    public Reservation deleteReservation(Integer id) {
        logger.traceEntry("Deleting reservation with id {}", id);
        try {
            Reservation reservation = reservationRepository.findOne(id).orElseThrow(() -> new ServiceException("Reservation not found"));
            Reservation result = reservationRepository.delete(id).orElseThrow(() -> new ServiceException("Failed to delete reservation"));
            logger.traceExit(result);
            return result;
        } catch (RepositoryException e) {
            logger.error("Error in deleteReservation: {}", e.getMessage());
            throw new ServiceException("Service error while deleting reservation", e);
        }
    }

    public List<Trip> findAllTrips(){
        logger.traceEntry("Finding all trips");
        try {
            Iterable<Trip> result = tripRepository.findAll();
            logger.traceExit(result);
            List<Trip> trips = new ArrayList<Trip>();
            result.forEach(trips::add);
            return trips;
        } catch (RepositoryException e) {
            logger.error("Error in findAllTrips: {}", e.getMessage());
            throw new ServiceException("Service error while finding all trips", e);
        }
    }


    public Iterable<Trip> searchTrips(){
        logger.traceEntry("Searching trips");
        try {
            Iterable<Trip> result = tripRepository.findAll();
            logger.traceExit(result);
            if(result == null) {
                throw new ServiceException("No trips found");
            }
            return result;
        } catch (RepositoryException e) {
            logger.error("Error in searchTrips: {}", e.getMessage());
            throw new ServiceException("Service error while searching trips", e);
        }
    }

    public List<Trip> searchTripsByDestTime(String destination, LocalDate dataDeparture, LocalTime startTime, LocalTime endTime){
        logger.traceEntry("Searching trips by destination and time");
        try {
            Iterable<Trip> result = tripRepository.findTripsByDestination(destination);
            logger.traceExit(result);
            if(result == null) {
                throw new ServiceException("No trips found");
            }
            List<Trip> trips = new ArrayList<Trip>();
            result.forEach(trips::add);
            List<Trip> tripsByTime = new ArrayList<Trip>();
            for(Trip trip : trips) {
                LocalTime tripStartTime = trip.getDepartureDate().toLocalTime();
                LocalDate date = trip.getDepartureDate().toLocalDate();
                if((tripStartTime.isAfter(startTime) && tripStartTime.isBefore(endTime)) && date == dataDeparture) {
                    tripsByTime.add(trip);
                }
            }
            return trips;
        } catch (RepositoryException e) {
            logger.error("Error in searchTripsByDestTime: {}", e.getMessage());
            throw new ServiceException("Service error while searching trips by destination and time", e);
        }
    }

    public List<Reservation> findAllReservations(){
        logger.traceEntry("Finding all reservations");
        try {
            Iterable<Reservation> result = reservationRepository.findAll();
            logger.traceExit(result);
            List<Reservation> reservations = new ArrayList<Reservation>();
            result.forEach(reservations::add);
            return reservations;
        } catch (RepositoryException e) {
            logger.error("Error in findAllReservations: {}", e.getMessage());
            throw new ServiceException("Service error while finding all reservations", e);
        }
    }

    @Override
    public void addObserver(Observer<TravelAgencyEventType> observer) {
        observers.add(observer);
    }

    @Override
    public void removeObserver(Observer<TravelAgencyEventType> observer) {
        observers.remove(observer);
    }

    @Override
    public void notifyObservers(TravelAgencyEventType e) {
        observers.stream().forEach(x -> x.update(e));
    }
}
